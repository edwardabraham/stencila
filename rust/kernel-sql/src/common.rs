use kernel::{
    common::itertools::Itertools,
    stencila_schema::{Datatable, ValidatorTypes},
};

/// Generate a `CREATE TABLE` statement for a Stencila [`Datatable`]
pub(crate) fn create_table_from_datatable(
    name: &str,
    datatable: &Datatable,
    or_replace: bool,
) -> String {
    let columns = datatable
        .columns
        .iter()
        .map(|column| {
            let item_validator = column
                .validator
                .as_deref()
                .and_then(|array_validator| array_validator.items_validator.clone());
            let data_type = match item_validator.as_deref() {
                Some(ValidatorTypes::BooleanValidator(..)) => "BOOLEAN",
                Some(ValidatorTypes::IntegerValidator(..)) => "INTEGER",
                Some(ValidatorTypes::NumberValidator(..)) => "REAL",
                Some(ValidatorTypes::StringValidator(..)) => "TEXT",
                Some(ValidatorTypes::DateValidator(..)) => "DATE",
                Some(ValidatorTypes::TimeValidator(..)) => "TIME",
                Some(ValidatorTypes::DateTimeValidator(..)) => "TIMESTAMP",
                Some(ValidatorTypes::TimestampValidator(..)) => "TIMESTAMP",
                Some(ValidatorTypes::DurationValidator(..)) => "INTERVAL",
                _ => "JSON",
            };

            let is_nullable = if let Some(validator) = &column.validator {
                if !validator.items_nullable {
                    "NOT NULL".to_string()
                } else {
                    String::new()
                }
            } else {
                String::new()
            };

            format!("{} {} {}", column.name, data_type, is_nullable)
        })
        .collect_vec()
        .join(", ");

    format!(
        "CREATE {} TABLE \"{}\"({});\n",
        if or_replace { "OR REPLACE" } else { "" },
        name,
        columns,
    )
}

/// Get the number of rows and columns of a Stencila [`Datatable`]
pub(crate) fn datatable_rows_cols(datatable: &Datatable) -> (usize, usize) {
    let rows = datatable
        .columns
        .first()
        .map(|column| column.values.len())
        .unwrap_or(0);

    let cols = datatable.columns.len();

    (rows, cols)
}
