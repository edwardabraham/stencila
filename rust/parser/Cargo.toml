[package]
name = "parser"
description = "Defines a specification and trait for parsers for semantically analyzing code"
version = "0.0.0"
edition = "2021"

[dependencies]
common = { path = "../common" }
formats = { path = "../formats" }
graph-triples = { path = "../graph-triples" }
