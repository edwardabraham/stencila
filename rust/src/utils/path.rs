use lexiclean::Lexiclean;
use std::path::{Path, PathBuf};

/// Merge paths
///
/// Differs from `path.join()` in that it joins the *parent* of `path1`, not `path1` itself.
/// Does not canonicalize the path, or check for its existence.
pub fn merge<P1: AsRef<Path>, P2: AsRef<Path>>(path1: P1, path2: P2) -> PathBuf {
    path1.as_ref().join("..").join(path2).lexiclean()
}

#[cfg(test)]
mod test {
    use super::*;
    use path_slash::PathBufExt;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_merge() {
        assert_eq!(merge("a", "c").to_slash_lossy(), "c");
        assert_eq!(merge("a/b", "c").to_slash_lossy(), "a/c");
        assert_eq!(merge("a/b/../d/e", "c").to_slash_lossy(), "a/d/c")
    }
}
