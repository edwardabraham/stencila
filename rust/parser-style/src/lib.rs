use std::path::Path;

use lightningcss::{
    stylesheet::{MinifyOptions, ParserOptions, PrinterOptions, StyleSheet},
    targets::Browsers,
    traits::ToCss,
};
use tailwind_css::TailwindBuilder;

use common::{
    eyre::{eyre, Result},
    itertools::Itertools,
    once_cell::sync::Lazy,
    regex::Regex,
};
use parser::{
    formats::Format,
    graph_triples::{
        relations,
        resources::{self, ResourceDigest},
        Resource, ResourceInfo,
    },
    utils::apply_tags,
    Parser, ParserTrait,
};

/// Regex for detecting document variables used in styles
pub static VAR_REGEX: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?:\$([a-zA-Z_][a-zA-Z_0-9]*)\b)|(?:\$\{\s*([a-zA-Z_][a-zA-Z_0-9]*)\s*\})")
        .expect("Unable to create regex")
});

/// A parser for the "Style" language
pub struct StyleParser;

impl ParserTrait for StyleParser {
    fn spec() -> Parser {
        Parser {
            language: Format::Style,
        }
    }

    fn parse(resource: Resource, path: &Path, code: &str) -> Result<ResourceInfo> {
        // The semantic content of the code (includes the language name and ignores comments)
        let mut semantics = Self::spec().language.to_string();
        let mut comments = Vec::new();
        let relations = code
            .split('\n')
            .enumerate()
            .fold(Vec::new(), |mut pairs, (row, line)| {
                // Skip the line if it is blank
                if line.trim().is_empty() {
                    return pairs;
                }

                // Parse comment line
                if line.trim_start().starts_with("//") {
                    comments.push((row, line));
                    return pairs;
                }

                let (start, expr) = (0, line);

                // Parse line for uses of variables
                for captures in VAR_REGEX.captures_iter(expr) {
                    let symbol = captures
                        .get(1)
                        .or_else(|| captures.get(2))
                        .expect("Should always have one group");
                    pairs.push((
                        relations::uses((row, start + symbol.start(), row, start + symbol.end())),
                        resources::symbol(path, symbol.as_str(), ""),
                    ))
                }

                // Add line to semantics
                semantics.push_str(line);
                semantics.push('\n');

                pairs
            });

        let mut resource_info = ResourceInfo::new(
            resource,
            Some(relations),
            None,
            None,
            Some(ResourceDigest::from_strings(code, Some(&semantics))),
            None,
            None,
        );

        // Apply tags from comments (this needs to be done at the end because tags
        // may remove pairs if `only` is specified)
        for (row, line) in comments {
            apply_tags(
                path,
                Self::spec().language,
                row,
                line,
                None,
                &mut resource_info,
            );
        }

        Ok(resource_info)
    }
}

/// Transpile a string of CSS or Tailwind to CSS
pub fn transpile_string(string: &str) -> Result<Vec<String>> {
    transpile_css(string)
        .or_else(|_| transpile_css_wrapped(string))
        .or_else(|_| transpile_tailwind(string))
}

/// Transpile Tailwind directives to CSS
fn transpile_tailwind(tw: &str) -> Result<Vec<String>> {
    // Transform Tailwind to CSS
    let mut tailwind = TailwindBuilder::default();
    let (.., css) = tailwind.inline(tw)?;

    // Parse the CSS
    transpile_css_wrapped(&css)
}

/// Wrap CSS in a `root` selector before transpiling
fn transpile_css_wrapped(css: &str) -> Result<Vec<String>> {
    transpile_css(&[":root {\n", css, "\n}"].concat())
}

/// Transpile CSS into a vector of CSS rules
fn transpile_css(css: &str) -> Result<Vec<String>> {
    let targets = Some(Browsers {
        ..Default::default()
    });

    // Parse the CSS into rules
    let mut sheet = StyleSheet::parse(
        css,
        ParserOptions {
            nesting: true,
            ..Default::default()
        },
    )
    .map_err(|error| eyre!("Error parsing CSS: {}\n{}", error.to_string(), css))?;

    // Optimize the rules
    sheet.minify(MinifyOptions {
        ..Default::default()
    })?;

    // Generate a vector of CSS rules
    let css = sheet
        .rules
        .0
        .iter()
        .filter_map(|rule| {
            rule.to_css_string(PrinterOptions {
                // It is important to specify targets here, even just defaults, so that nested rules
                // get expanded
                targets,
                // Do not minify during development so debugging easier and snapshots easier to read
                minify: !cfg!(debug_assertions),
                ..Default::default()
            })
            .ok()
        })
        .collect_vec();

    Ok(css)
}

#[cfg(test)]
mod tests {
    use test_snaps::{
        insta::{assert_json_snapshot, assert_snapshot},
        snapshot_fixtures,
    };
    use test_utils::fixtures;
    use test_utils::pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn parse_style_fragments() {
        snapshot_fixtures("fragments/style/*.style", |path| {
            let code = std::fs::read_to_string(path).expect("Unable to read");
            let path = path.strip_prefix(fixtures()).expect("Unable to strip");
            let resource = resources::code(path, "", "SoftwareSourceCode", Format::SQL);
            let resource_info = StyleParser::parse(resource, path, &code).expect("Unable to parse");
            assert_json_snapshot!(resource_info);
        })
    }

    #[test]
    fn test_transpile_css() -> Result<()> {
        // Nesting
        assert_snapshot!(transpile_css(
            r#"
            .foo {
                color: pink;
                & .bar {
                    color: red;
                    & .baz .quax {
                        color: orange;
                    }
                }
            }
            "#,
        )?
        .join("\n\n"));

        // Media queries
        assert_snapshot!(transpile_css(
            r#"
            @media (min-width: 640px) {
                :root {
                    font-size: 1rem;
                    line-height: 1.5rem;
                }
            }
            "#,
        )?
        .join("\n\n"));

        Ok(())
    }

    #[test]
    fn test_transpile_tailwind() -> Result<()> {
        // Basic
        assert_snapshot!(transpile_tailwind(r"text-md text-red-200 bg-red-100")?[0]);

        // Tests of support for Tailwind classes by `tailwind-css`
        // Border radius
        assert_snapshot!(transpile_tailwind(r"rounded(lg")?[0]);
        // Border width
        assert_snapshot!(transpile_tailwind(r"border(2 x-4 t-8 b-0)")?[0]);
        // Border color
        assert_snapshot!(transpile_tailwind(r"border(rose-400)")?[0]);
        // Border style
        assert_snapshot!(transpile_tailwind(r"border(dashed)")?[0]);

        Ok(())
    }

    /// Test Tailwind variants e.g. sm, md, lg and hover
    ///
    /// Currently, handling of variants is not implemented by `tailwind-css`. These tests
    /// therefore only produce the CSS for the last variant. The snapshots are expected
    /// to break when this is implemented.
    #[test]
    fn test_transpile_tailwind_variants() -> Result<()> {
        // Four ways to do same thing with breakpoints
        assert_snapshot!(
            transpile_tailwind(r"sm:text-md sm:text(md) sm:(text(md)) sm:(text-md)")?.join("\n\n")
        );

        // As above but using different sizes
        assert_snapshot!(transpile_tailwind(
            r"sm:text-md md:text(lg) lg:(text(xl)) xl:(text-5xl)"
        )?
        .join("\n\n"));

        Ok(())
    }

    #[test]
    fn test_transpile_string() -> Result<()> {
        let css1 = transpile_string(r"font-size: 1rem; line-height: 1.5rem")?;
        let css2 = transpile_string(r"text-md")?;
        assert_eq!(css1, css2);

        Ok(())
    }
}
