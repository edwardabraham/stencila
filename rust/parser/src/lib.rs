use std::path::Path;

use common::{
    eyre::Result,
    serde::{Deserialize, Serialize},
    serde_with::skip_serializing_none,
};
use formats::Format;
use stencila_schema::{ExecutionAuto, ExecutionDependency, ExecutionDependent, ExecutionTag};

mod files;
pub use files::*;
mod tags;
pub use tags::*;
mod variables;
pub use variables::*;

// Export and re-export for the convenience of crates that implement a parser
pub mod utils;
pub use common;
pub use formats;
pub use hash_utils;
pub use stencila_schema;

/// A specification for parsers
///
/// All parsers, including those implemented in plugins, should provide this
/// specification. Rust implementations return a `Parser` instance from the
/// `spec` function of `ParserTrait`. Plugins provide a JSON or YAML serialization
/// as part of their manifest.
#[derive(Debug, Default, Clone, Serialize, Deserialize)]
#[serde(default, rename_all = "camelCase", crate = "common::serde")]
pub struct Parser {
    /// The language that the parser parses
    pub language: Format,
}

/// A trait for parsers
///
/// Defines similar functions to `serde_json` (and other `serde_` crates) for
/// converting nodes to/from strings, files, readers etc.
pub trait ParserTrait {
    /// Get the [`Parser`] specification
    fn spec() -> Parser;

    /// Parse some code into a [`ParseInfo`] object
    ///
    /// # Arguments
    ///
    /// - `code`: The code to parse
    /// - `path`: The filesystem path of the code. Used to resolve relative
    ///           file paths in parsed code.
    fn parse(code: &str, path: Option<&Path>) -> Result<ParseInfo>;
}

#[skip_serializing_none]
#[derive(Debug, Default, Serialize)]
#[serde(default, rename_all = "camelCase", crate = "common::serde")]
pub struct ParseInfo {
    /// Whether the code has any syntax errors
    pub syntax_errors: bool,

    /// The semantic digest of the code
    pub semantic_digest: u64,

    /// The execution dependencies of the code
    pub execution_dependencies: Vec<ExecutionDependency>,

    /// The execution dependents of the code
    pub execution_dependents: Vec<ExecutionDependent>,

    /// The execution tags that were parsed
    pub execution_tags: Vec<ExecutionTag>,

    /// Whether the code has been tagged as pure or not
    pub execution_pure: Option<bool>,

    /// Whether the code has been tagged with an autorun option
    pub execution_auto: Option<ExecutionAuto>,
}
