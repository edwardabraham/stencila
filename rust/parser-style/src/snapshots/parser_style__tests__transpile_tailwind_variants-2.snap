---
source: rust/parser-style/src/lib.rs
assertion_line: 249
expression: "transpile_tailwind(r\"sm:text-md md:text(lg) lg:(text(xl)) xl:(text-5xl)\")?.join(\"\\n\\n\")"
---
:root {
  font-size: 3rem;
  line-height: 100%;
}
